dbms/src/Flash/Mpp/ReceivedMessageQueue.h

```diff
@@ -55,6 +57,65 @@ enum class ReceiverMode
     Async
 };
 
+class GRPCNotifyQueue : public NotifyFuture
+{
+public:
+    template <typename... Args>
+    explicit GRPCNotifyQueue(const LoggerPtr & log_, Args &&... args)
+        : queue(log_, std::forward<Args>(args)...)
+    {}
+
+    void registerTask(TaskPtr && task) override { queue.registerPipeReadTask(std::move(task)); }
+
+    MPMCQueueResult pop(ReceivedMessagePtr & data) { return queue.pop(data); }
+
+    MPMCQueueResult tryPop(ReceivedMessagePtr & data) { return queue.tryPop(data); }
+
+    MPMCQueueResult forcePush(ReceivedMessagePtr && data) { return queue.forcePush(std::move(data)); }
+
+    MPMCQueueResult push(ReceivedMessagePtr && data) { return queue.push(std::move(data)); }
+
+    MPMCQueueResult tryDequeue() { return queue.tryDequeue(); }
+
+    MPMCQueueResult pushWithTag(ReceivedMessagePtr && data, GRPCKickTag * new_tag)
+    {
+        return queue.pushWithTag(std::move(data), new_tag);
+    }
+
+    void setKickFuncForTest(GRPCKickFunc && func) { queue.setKickFuncForTest(std::move(func)); }
+
+    bool finish() { return queue.finish(); }
+    bool cancel() { return queue.cancel(); }
+
+    bool isWritable() const { return queue.isWritable(); }
+
+    void registerPipeWriteTask(TaskPtr && task) { queue.registerPipeWriteTask(std::move(task)); }
+
+private:
+    GRPCRecvQueue<ReceivedMessagePtr> queue;
+};
+
+class MSGChannel : public NotifyFuture
+{
+public:
+    void registerTask(TaskPtr && task) override { queue_ref.registerPipeReadTask(std::move(task)); }
+
+    MPMCQueueResult pop(ReceivedMessagePtr & data) { return queue_ref.pop(data); }
+
+    MPMCQueueResult tryPop(ReceivedMessagePtr & data) { return queue_ref.tryPop(data); }
+
+    MPMCQueueResult forcePush(const ReceivedMessagePtr & data) { return queue_ref.forcePush(data); }
+
+    bool finish() { return queue_ref.finish(); }
+    bool cancel() { return queue_ref.cancel(); }
+
+private:
+    using QueueImpl = LooseBoundedMPMCQueue<ReceivedMessagePtr>;
+    // these are unbounded queues.
+    std::unique_ptr<QueueImpl> queue = std::make_unique<QueueImpl>(std::numeric_limits<size_t>::max());
```

Comment:
Why use `unique_ptr`? Also, how about letting `MSGChannel` inherits `LooseBoundedMPMCQueue<ReceivedMessagePtr>`?

Reply:
ref https://github.com/pingcap/tiflash/pull/7963,
Because LooseBoundedMPMCQueue includes locks, it must use unique_ptr or shared_ptr somewhere.